core/tests/protractor/libraryFlow.js

```diff
@@ -0,0 +1,224 @@
+  });
+
+  it('visits the top rated page', function() {
+    // To visit the top rated page, at least one
+    // exploration has to be rated by the user
+    if (browser.isMobile) {
+      rateExplorationInMobile();
+    } else {
+      rateExplorationInDesktop();
+    }
+    libraryPage.get();
+    element(by.css('.protractor-test-library-top-rated')).click();
+    waitFor.pageToFullyLoad();
+    expect(browser.getCurrentUrl()).toContain('library/top_rated');
+  });
+
+  it('visits the recent explorations page', function() {
+    visitRecentlyPublishedPage();
+    expect(browser.getCurrentUrl()).toContain('library/recently_published');
+  });
+
+  afterEach(function() {
+    if (browser.isMobile) {
+      general.checkForConsoleErrors([
+        // TODO(apb7): Remove these when https://github.com/oppia/oppia/issues/5363 is resolved.
+        'http://localhost:9001/third_party/static/angularjs-1.5.8/angular.min.js 117:9 "Mismatch"',
+        'http://localhost:9001/third_party/static/angularjs-1.5.8/angular.min.js 117:9 "SearchQuery: Test Exploration"',
+        'http://localhost:9001/third_party/static/angularjs-1.5.8/angular.min.js 117:9 "Input: "',
+        'http://localhost:9001/third_party/static/angularjs-1.5.8/angular.min.js 117:9 "Mismatch"',
+        'http://localhost:9001/third_party/static/angularjs-1.5.8/angular.min.js 117:9 "SearchQuery: python"',
+        'http://localhost:9001/third_party/static/angularjs-1.5.8/angular.min.js 117:9 "Input: "'
+      ]);
+    } else {
+      general.checkForConsoleErrors([]);
+    }
+  });
+});
+
+describe('Rating', function() {
+  var EXPLORATION_RATINGTEST = 'RatingTest';
+  var CATEGORY_BUSINESS = 'Business';
+  var LANGUAGE_ENGLISH = 'English';
+  var MINIMUM_ACCEPTABLE_NUMBER_OF_RATINGS = 1;
+  var adminPage = null;
+  var libraryPage = null;
+  var explorationPlayerPage = null;
+
+  beforeEach(function() {
+    adminPage = new AdminPage.AdminPage();
+    libraryPage = new LibraryPage.LibraryPage();
+    explorationPlayerPage = new ExplorationPlayerPage.ExplorationPlayerPage();
+  });
+
+  var addRating = function(userEmail, userName, explorationName, ratingValue) {
+    users.createUser(userEmail, userName);
+    users.login(userEmail);
+    libraryPage.get();
+    libraryPage.findExploration(EXPLORATION_RATINGTEST);
+    libraryPage.playExploration(EXPLORATION_RATINGTEST);
+
+    // The exploration header is only visible in desktop browsers.
+    if (!browser.isMobile) {
+      explorationPlayerPage.expectExplorationNameToBe(explorationName);
+    }
+    explorationPlayerPage.rateExploration(ratingValue);
+    users.logout();
+  };
+
+  it('should display ratings on exploration when minimum ratings have been ' +
+     'submitted', function() {
+    users.createUser('user11@explorationRating.com', 'user11Rating');
+    // Create a test exploration.
+    users.login('user11@explorationRating.com', true);
+
+    // We need a test exploration here.
+    if (browser.isMobile) {
```

Comment:
Btw, this is a good pattern for this sort of "setup" thing -- i.e., limit the branching to a single call that's different for mobile + desktop. Nice! You can replicate it in other places.

Reply:
Sure, sounds good! Thanks!

Reply:
So ... I want to draw your attention again to this comment, which was trying to highlight that it was good for the mobile/desktop differences to be minimal and single-line. In addition, this example shows why it's really clear that the behaviour has to diverge for both, while also making clear that both branches basically have the same effect. As you can see in this example, it's clear that in both branches the exploration being created is basically the same ("rating test") and there is no difference at all in terms of how you are finding it in the library page. (In general try to get rid of branching entirely, where possible.)

This is not the case for most of the other if/else branches I'm seeing. It would be better to make them more like this case, and minimize the branching. Often, you should be able to make minor modifications to the existing tests to ensure parity.

I'm going to stop reviewing here; my comments in general apply throughout. I feel like, once you get this right, then any remaining issues should be trivial to clean up.